domains/client/domain-operator/src/snap_sync.rs

```diff
@@ -0,0 +1,481 @@
+where
+    NR: NetworkRequest + Send + Sync,
+    CNR: NetworkRequest + Send + Sync + 'static,
+    Block: BlockT,
+    CBlock: BlockT,
+    AS: AuxStore,
+{
+    pub domain_client: Arc<DomainClient>,
+    pub sync_service: Arc<SyncingService<Block>>,
+    pub domain_fork_id: Option<String>,
+    pub domain_network_request: NR,
+    pub consensus_client: Arc<CClient>,
+    pub domain_block_downloader: Arc<dyn BlockDownloader<Block>>,
+    pub consensus_chain_sync_params: ConsensusChainSyncParams<Block, CBlock, CNR, AS>,
+}
+
+async fn get_last_confirmed_block<Block: BlockT>(
+    block_downloader: Arc<dyn BlockDownloader<Block>>,
+    sync_service: &SyncingService<Block>,
+    block_number: BlockNumber,
+) -> Result<BlockData<Block>, sp_blockchain::Error> {
+    const LAST_CONFIRMED_BLOCK_RETRIES: u32 = 5;
+    const LOOP_PAUSE: Duration = Duration::from_secs(20);
+    const MAX_GET_PEERS_ATTEMPT_NUMBER: usize = 30;
+
+    for attempt in 1..=LAST_CONFIRMED_BLOCK_RETRIES {
+        debug!(%attempt, %block_number, "Starting last confirmed block request...");
+
+        debug!(%block_number, "Gathering peers for last confirmed block request.");
+        let mut tried_peers = HashSet::<PeerId>::new();
+
+        let mut current_get_peers_attempt = 0usize;
+        let current_peer_id = loop {
+            current_get_peers_attempt += 1;
+
+            if current_get_peers_attempt > MAX_GET_PEERS_ATTEMPT_NUMBER {
+                return Err(sp_blockchain::Error::Application(
+                    "Got empty state blocks collection for domain snap sync".into(),
+                ));
+            }
+            let all_connected_peers = sync_service
+                .peers_info()
+                .await
+                .expect("Network service must be available.");
+
+            debug!(
+                %current_get_peers_attempt,
+                ?all_connected_peers,
+                "Get last confirmed block: all connected peers"
+            );
+
+            let connected_full_peers = all_connected_peers
+                .iter()
+                .filter_map(|(peer_id, info)| (info.roles.is_full()).then_some(*peer_id))
+                .collect::<Vec<_>>();
+
+            debug!(
+                %current_get_peers_attempt,
+                ?tried_peers,
+                "Get last confirmed block: sync peers: {:?}", connected_full_peers
+            );
+
+            let active_peers_set = HashSet::from_iter(connected_full_peers.into_iter());
+
+            if let Some(peer_id) = active_peers_set.difference(&tried_peers).next().cloned() {
+                break peer_id;
+            }
+
+            sleep(LOOP_PAUSE).await;
+        };
+
+        tried_peers.insert(current_peer_id);
+
+        let id = {
+            let now = SystemTime::now();
+            let duration_since_epoch = now
+                .duration_since(UNIX_EPOCH)
+                .expect("Time usually goes forward");
```

Comment:
These are both rare situations, but they might be worth tweaking the error messages a bit:

1. This expect will panic if the system clock is before 1970, which can happen in rare cases if the user makes a typo in the year (particularly Windows), if the clock battery is flat, or the VM time is misconfigured. So we might want to make the panic message more explicit, or turn it into a fatal error:
```suggestion
                .expect("Subspace needs a clock set to 2021 or later");
```

2. Similarly, I’ve seen a few VMs with misconfigured time virtualisation, that always return the same time. How important is it that the ID is unique?

Reply:
It doesn't seem to be an issue because the request-response sequence should get several orders of magnitude more time than the possible "repeated ID". There are other reasons: we have tolerance level of repeated IDs = 2, we need only the first successful response that will be the case in the absolute majority of cases with successfully established connections, and finally, we'll likely replace this particular snippet in favor of local block generation in the further updates. 

Anyway, I appreciate the detailed explanation of possible issues with the clock!